src/python/pants/build_graph/target.py

```diff
@@ -473,7 +473,10 @@ def transitive_invalidation_hash(self, fingerprint_strategy=None, depth=0):
       def dep_hash_iter():
         for dep in self.dependencies:
           try:
-            dep_hash = dep.transitive_invalidation_hash(fingerprint_strategy, depth=depth+1)
+            if getattr(self, 'strict_deps', None):
```

Comment:
Sorry for changing my mind again. Implement this logic in fingerprintstrategy is a bit tricky. target.invalidation_hash and target.transitive_invalidation_hash both use fingerprintstrategy, thus will result in recursive invalidation in unwanted case. I think it's cleaner to modify key_for_target method.